NetKet/Machine/mps_diagonal.hpp

```diff
@@ -0,0 +1,471 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// by S. Efthymiou, October 2018
+
+#include <Eigen/Dense>
+#include <iostream>
+#include <vector>
+#include "Lookup/lookup.hpp"
+#include "Utils/all_utils.hpp"
+
+#ifndef NETKET_MPS_DIAGONAL_HPP
+#define NETKET_MPS_DIAGONAL_HPP
+
+namespace netket {
+
+template <typename T>
+class MPSDiagonal : public AbstractMachine<T> {
+  using VectorType = typename AbstractMachine<T>::VectorType;
+  // using MatrixType = typename AbstractMachine<T>::MatrixType;
+
+  // Number of sites
+  int N_;
+  // Physical dimension
+  int d_;
+  // Bond dimension
+  int D_;
+  // Number of variational parameters
+  int npar_;
+  // Period of translational symmetry (has to be a divisor of N)
+  int symperiod_;
+
+  // MPS Matrices (stored as [symperiod, d, D, D]
+  std::vector<std::vector<VectorType>> W_;
+
+  // Map from Hilbert states to MPS indices
+  std::map<double, int> confindex_;
+
+  const Hilbert &hilbert_;
+
+ public:
+  using StateType = T;
+  using LookupType = Lookup<T>;
+
+  // constructor as a machine
+  explicit MPSDiagonal(const Hilbert &hilbert, const json &pars)
+      : N_(hilbert.Size()), d_(hilbert.LocalSize()), hilbert_(hilbert) {
+    from_json(pars);
+  }
+
+  // Auxiliary function that defines the matrices
+  void Init() {
+    // Initialize parameters
+    std::vector<VectorType> pushback_vec;
+    VectorType init_mat = VectorType::Zero(D_);
+    npar_ = symperiod_ * d_ * D_;
+
+    for (int site = 0; site < symperiod_; site++) {
+      W_.push_back(pushback_vec);
+      for (int spin = 0; spin < d_; spin++) {
+        W_[site].push_back(init_mat);
+      }
+    }
+
+    // Machine creation messages
+
+    InfoMessage() << "Periodic diagonal MPS machine with " << N_
+                  << " sites created" << std::endl;
+    InfoMessage() << "Physical dimension d = " << d_
+                  << " and bond dimension D = " << D_ << std::endl;
+    if (symperiod_ < N_) {
+      InfoMessage() << "Translation invariance is used. Number of "
+                       "variational parameters is "
+                    << npar_ << " instead of " << npar_ * N_ / symperiod_
+                    << std::endl;
+    } else {
+      InfoMessage() << "Number of variational parameters is " << npar_
+                    << std::endl;
+    }
+
+    // Initialize map from Hilbert space states to MPS indices
+    auto localstates = hilbert_.LocalStates();
+    for (int i = 0; i < d_; i++) {
+      confindex_[localstates[i]] = i;
+    }
+  }
+
+  int Npar() const override { return npar_; }
+
+  VectorType GetParameters() override {
+    int k = 0;
+    VectorType pars(npar_);
+
+    for (int site = 0; site < symperiod_; site++) {
+      for (int spin = 0; spin < d_; spin++) {
+        for (int i = 0; i < D_; i++) {
+          pars(k) = W_[site][spin](i);
```

Comment:
Eigen has also a (reasonably mature) [Tensor type](http://eigen.tuxfamily.org/index.php?title=Tensor_support), maybe it might be convenient to work with that to store the W?